kong/tools/channel.lua

```diff
@@ -0,0 +1,164 @@
+end
+
+--- waiting version of `d:add()`
+--- blocks the coroutine until there's no value under this key
+--- so the new value can be safely added
+local function add_wait(dict, key, val, exptime, deadline)
+  local step = 0
+
+  while deadline > now() do
+    local ok, err = dict:add(key, val, exptime)
+    if ok then
+      return true
+    end
+
+    if err ~= "exists" then
+      return nil, err
+    end
+
+    step = waitstep(step, deadline)
+  end
+
+  return nil, "timeout"
+end
+
+--- waiting version of `d:get()`
+--- blocks the coroutine until there's actually a value under this key
+local function get_wait(dict, key, deadline)
+  local step = 0
+
+  while deadline > now() do
+    local value, err = dict:get(key)
+    if value then
+      return value
+    end
+
+    if err ~= nil then
+      return nil, err
+    end
+
+    step = waitstep(step, deadline)
+  end
+
+  return nil, "timeout"
+end
+
+--- waits until the key is empty
+--- blocks the coroutine while there's a value under this key
+local function empty_wait(dict, key, deadline)
+  local step = 0
+  while deadline > now() do
+    local value, err = dict:get(key)
+    if not value then
+      if err ~= nil then
+        return nil, err
+      end
+
+      return true
+    end
+
+    step = waitstep(step, deadline)
+  end
+  return nil, "timeout"
+end
+
+
+local Channel = {}
+Channel.__index = Channel
+
+--- Create a new channel client
+--- @param dict_name string Name of the shdict to use
+--- @param name string channel name
+function Channel.new(dict_name, name)
+  return setmetatable({
+    dict = assert(ngx.shared[dict_name]),
+    name = name,
+    exptime = DEFAULT_EXPTIME,
+    timeout = DEFAULT_TIMEOUT,
+  }, Channel)
+end
+
+
+--- Post a value, client -> server
+--- blocks the thread until the server picks it
+--- @param val any Value to post (any type supported by shdict)
+--- @return boolean, string ok, err
+function Channel:post(val)
+  local key = POST_VAL_KEY_PREFIX .. self.name
+  local ok, err = add_wait(self.dict, key, val, self.exptime, now() + self.timeout)
+  if not ok then
+    return nil, err
+  end
+
+  ok, err = add_wait(self.dict, NAME_KEY, self.name, self.exptime, now() + self.timeout)
+  if not ok then
+    self.dict:delete(key)
+    return nil, err
+  end
+
+  return empty_wait(self.dict, key, now() + self.timeout)
+end
+
+--- Get a response value, client <- server
+--- blocks the thread until the server puts a value
+--- @return any, string value, error
+function Channel:get()
+  local key = RESP_VAL_KEY_PREFIX .. self.name
+  local val, err = get_wait(self.dict, key, now() + self.timeout)
+  if val then
+    self.dict:delete(key)
+    return val
+  end
+
+  return nil, err
+end
+
+
+--- Waits until a value is posted by any client
+--- @param dict shdict shdict to use
+--- @return any, string, string value, channel name, error
+function Channel.wait_all(dict)
+  local name, err = get_wait(dict, NAME_KEY, now() + DEFAULT_TIMEOUT)
```

Comment:
Could we just use one dictionary key, and put the name of the channel together into the value? So we could fetch only once.
After all, post keeps blocked if `NAME_KEY` is not free.

Reply:
this is one of the reasons to move to a socket implementation, it's hard to follow which concepts tie 1:1 into which and which are 1:n
in this case, i think the purpose was to reduce serialization of processes, each client pushes their value without blocking each other, and only block on making the "server" pull the notification.    at higher level i think there's still the same contention anyway... but it's important to be able for each client to fire the message and not wait for all the others to be served.